baselines/pfsp/pfsp_multigpu_cuda.c

```diff
@@ -0,0 +1,602 @@
+    int *machine_pair_order_d;
+
+    // Allocating and copying memory necessary for deep copy of lbound2
+    int nb_mac_pairs = lbound2->nb_machine_pairs;
+    cudaMalloc((void**)&johnson_schedule_d, nb_mac_pairs * jobs * sizeof(int));
+    cudaMalloc((void**)&lags_d, nb_mac_pairs * jobs * sizeof(int));
+    cudaMalloc((void**)&machine_pairs_1_d, nb_mac_pairs * sizeof(int));
+    cudaMalloc((void**)&machine_pairs_2_d, nb_mac_pairs * sizeof(int));
+    cudaMalloc((void**)&machine_pair_order_d, nb_mac_pairs * sizeof(int));
+    cudaMemcpy(johnson_schedule_d, lbound2->johnson_schedules, nb_mac_pairs * jobs * sizeof(int), cudaMemcpyHostToDevice);
+    cudaMemcpy(lags_d, lbound2->lags, nb_mac_pairs * jobs * sizeof(int), cudaMemcpyHostToDevice);
+    cudaMemcpy(machine_pairs_1_d, lbound2->machine_pairs_1, nb_mac_pairs * sizeof(int), cudaMemcpyHostToDevice);
+    cudaMemcpy(machine_pairs_2_d, lbound2->machine_pairs_2, nb_mac_pairs * sizeof(int), cudaMemcpyHostToDevice);
+    cudaMemcpy(machine_pair_order_d, lbound2->machine_pair_order, nb_mac_pairs * sizeof(int), cudaMemcpyHostToDevice);
+
+    // Deep copy of lbound2
+    lbound2_d.johnson_schedules = johnson_schedule_d;
+    lbound2_d.lags = lags_d;
+    lbound2_d.machine_pairs_1 = machine_pairs_1_d;
+    lbound2_d.machine_pairs_2 = machine_pairs_2_d;
+    lbound2_d.machine_pair_order = machine_pair_order_d;
+    lbound2_d.nb_machine_pairs = lbound2->nb_machine_pairs;
+    lbound2_d.nb_jobs = lbound2->nb_jobs;
+    lbound2_d.nb_machines = lbound2->nb_machines;
+
+    // Local pools retrieving their respective works from pool
+    unsigned long long int tree = 0, sol = 0;
+    SinglePool pool_loc;
+    initSinglePool(&pool_loc);
+
+    for (int i = 0; i < c; i++) {
+      pool_loc.elements[i] = pool.elements[gpuID+f+i*D];
+    }
+    pool_loc.size += c;
+    if (gpuID == D-1) {
+      for (int i = c; i < l; i++) {
+        pool_loc.elements[i] = pool.elements[(D*c)+f+i-c];
+      }
+      pool_loc.size += l-c;
+    }
+    
+    // Allocating parents vector on CPU and GPU
+    Node* parents = (Node*)malloc(M * sizeof(Node));
+    Node* parents_d;
+    cudaMalloc((void**)&parents_d, M * sizeof(Node));
+    
+    // Allocating bounds vector on CPU and GPU
+    int* bounds = (int*)malloc((jobs*M) * sizeof(int));
+    int *bounds_d;
+    cudaMalloc((void**)&bounds_d, (jobs*M) * sizeof(int));
+    
+      
+    while (1) {
+      int poolSize = pool_loc.size;
+      if (poolSize >= m) {
+	poolSize = MIN(poolSize,M);
+      
+	for(int i= 0; i < poolSize; i++) {
+	  int hasWork = 0;
+	  parents[i] = popBack(&pool_loc,&hasWork);
+	  if (!hasWork) break;
+	}
+	
+	/*
+	  TODO: Optimize 'numBounds' based on the fact that the maximum number of
+	  generated children for a parent is 'parent.limit2 - parent.limit1 + 1' or
+	  something like that.
+	*/
+	const int numBounds = jobs * poolSize;   
+	const int nbBlocks = ceil((double)numBounds / BLOCK_SIZE);
+	const int nbBlocks_lb1_d = ceil((double)nbBlocks/jobs); 
```

Comment:
I think this computation could be done directly in `evaluate_gpu` and only if we choose `lb1_d`. Doing this, we can avoid this useless (most of the time) computation **each iteration** and remove one argument in the `evaluate_gpu` function. Do you agree?

Reply:
This is also true for the single-GPU version

Reply:
It's better to let it as it is for the moment, but add a TODO

Reply:
Fixed. I still had to declare `int nbBlocks_lb1_d;` outside of the switch environment as he would give errors trying to declare new variables inside just one of the switch clauses. For the computation ` nbBlocks_lb1_d = ceil((double)nbBlocks/jobs);` is being done only foor **lb1_d** now. Check `evaluate.cu` file.